src/VisualStudio/Core/Def/Implementation/Snippets/AbstractSnippetExpansionClient.cs

```diff
@@ -352,13 +352,14 @@ public virtual bool TryHandleEscape()
 
         public virtual bool TryHandleReturn()
         {
-            // TODO(davip): Only move the caret if the enter was hit within the editable spans
-
             if (ExpansionSession != null)
             {
-                ExpansionSession.EndCurrentExpansion(fLeaveCaret: 0);
+                // Only move the caret if the enter was hit within the snippet fields.
+                var hitWithinField = VSConstants.S_OK == ExpansionSession.GoToNextExpansionField(fCommitIfLast: 0);
```

Comment:
I did not examine the value myself, code was copied and modified from `TryHandleTab` if you expand the diff upwards a few times. The [document](https://msdn.microsoft.com/en-us/library/microsoft.visualstudio.textmanager.interop.ivsexpansionsession.gotonextexpansionfield.aspx) says it would be an error value.